File: infer/tests/codetoanalyze/objc/frontend/boxing/array.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <Foundation/Foundation.h>

int main() {

  NSString* s;

  NSArray* germanCars = @[
    @"Mercedes-Benz",
    @"BMW",
    @"Porsche",
    @"Opel",
    @"Volkswagen",
    @"Audi"
  ];
  s = germanCars[3];

  for (NSString* item in germanCars) {
    NSLog(@"%@", item);
  }

  return 0;
}

File: infer/tests/codetoanalyze/c/purity/array.c
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <stdio.h>
#include <time.h>

void swap_bad(int* array, int i, int j) {
  int tmp = array[i];
  array[i] = array[j];
  array[j] = tmp;
}

void alias_mod_bad(int array[], int i, int j) {
  int* a = array;
  a[j] = i;
}

void fresh_arr_ok(int size) {
  int arr[size];
  for (int i = 0; i < size - 1; i++) {
    arr[i] = 0;
  }
}

void call_impure_with_local_ok(int size) {
  int arr[size];
  alias_mod_bad(arr, 0, 9);
}

void time_bad() {
  time_t rawtime;
  struct tm* timeinfo;

  time(&rawtime);
  timeinfo = localtime(&rawtime);
}
